File: gitb-ui/app/controllers/SystemService.scala
/*
 * Copyright (C) 2025 European Union
 *
 * Licensed under the EUPL, Version 1.2 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence"); You may not use this work except in compliance with the Licence.
 *
 * You may obtain a copy of the Licence at:
 *
 * https://interoperable-europe.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on an
 * "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the Licence for
 * the specific language governing permissions and limitations under the Licence.
 */

package controllers

import config.Configurations
import controllers.util._
import exceptions._
import managers._
import models.Systems
import models.prerequisites.PrerequisiteUtil
import org.apache.commons.io.FileUtils
import play.api.libs.json.Json
import play.api.mvc._
import utils.{JsonUtil, RepositoryUtils}

import javax.inject.Inject
import scala.concurrent.{ExecutionContext, Future}

class SystemService @Inject() (repositoryUtils: RepositoryUtils,
                               testResultManager: TestResultManager,
                               authorizedAction: AuthorizedAction, cc:
                               ControllerComponents, systemManager: SystemManager,
                               authorizationManager: AuthorizationManager,
                               communityLabelManager: CommunityLabelManager)
                              (implicit ec: ExecutionContext) extends AbstractController(cc) {

  def deleteSystem(systemId:Long): Action[AnyContent] = authorizedAction.async { request =>
    authorizationManager.canDeleteSystem(request, systemId).flatMap { _ =>
      systemManager.deleteSystemWrapper(systemId).map { _ =>
        ResponseConstructor.constructEmptyResponse
      }
    }
  }

  def registerSystemWithOrganization: Action[AnyContent] = authorizedAction.async { request =>
    val paramMap = ParameterExtractor.paramMap(request)
    val userId = ParameterExtractor.extractUserId(request)
    val system:Systems = ParameterExtractor.extractSystemWithOrganizationInfo(paramMap)
    authorizationManager.canCreateSystem(request, system.owner).flatMap { _ =>
      val otherSystem = ParameterExtractor.optionalLongBodyParameter(paramMap, Parameters.OTHER_SYSTEM)
      val copySystemParameters = ParameterExtractor.optionalBodyParameter(paramMap, Parameters.SYSTEM_PARAMETERS).getOrElse("false").toBoolean
      val copyStatementParameters = ParameterExtractor.optionalBodyParameter(paramMap, Parameters.STATEMENT_PARAMETERS).getOrElse("false").toBoolean
      val values = ParameterExtractor.extractSystemParameterValues(paramMap, Parameters.PROPERTIES, optional = true)
      val files = ParameterExtractor.extractFiles(request).map {
        case (key, value) => (key.substring(key.indexOf('_')+1).toLong, value)
      }
      if (Configurations.ANTIVIRUS_SERVER_ENABLED && ParameterExtractor.virusPresentInFiles(files.map(entry => entry._2.file))) {
        Future.successful {
          ResponseConstructor.constructBadRequestResponse(ErrorCodes.VIRUS_FOUND, "File failed virus scan.")
        }
      } else {
        systemManager.registerSystemWrapper(userId, system, otherSystem, values, Some(files), copySystemParameters, copyStatementParameters).map { _ =>
          ResponseConstructor.constructEmptyResponse
        }
      }
    }.andThen { _ =>
      if (request.body.asMultipartFormData.isDefined) request.body.asMultipartFormData.get.files.foreach { file => FileUtils.deleteQuietly(file.ref) }
    }
  }

  /**
   * Updates the profile of a system
   */
  def updateSystemProfile(sutId:Long): Action[AnyContent] = authorizedAction.async { request =>
    authorizationManager.canUpdateSystem(request, sutId).flatMap { _ =>
      val paramMap = ParameterExtractor.paramMap(request)
      val sname = ParameterExtractor.requiredBodyParameter(paramMap, Parameters.SYSTEM_SNAME)
      val fname = ParameterExtractor.requiredBodyParameter(paramMap, Parameters.SYSTEM_FNAME)
      val descr = ParameterExtractor.optionalBodyParameter(paramMap, Parameters.SYSTEM_DESC)
      val version = ParameterExtractor.optionalBodyParameter(paramMap, Parameters.SYSTEM_VERSION)
      val otherSystem = ParameterExtractor.optionalLongBodyParameter(paramMap, Parameters.OTHER_SYSTEM)
      val copySystemParameters = ParameterExtractor.optionalBodyParameter(paramMap, Parameters.SYSTEM_PARAMETERS).getOrElse("false").toBoolean
      val copyStatementParameters = ParameterExtractor.optionalBodyParameter(paramMap, Parameters.STATEMENT_PARAMETERS).getOrElse("false").toBoolean
      val values = ParameterExtractor.extractSystemParameterValues(paramMap, Parameters.PROPERTIES, optional = true)
      val files = ParameterExtractor.extractFiles(request).map {
        case (key, value) => (key.substring(key.indexOf('_')+1).toLong, value)
      }
      if (Configurations.ANTIVIRUS_SERVER_ENABLED && ParameterExtractor.virusPresentInFiles(files.map(entry => entry._2.file))) {
        Future.successful {
          ResponseConstructor.constructBadRequestResponse(ErrorCodes.VIRUS_FOUND, "File failed virus scan.")
        }
      } else {
        val userId = ParameterExtractor.extractUserId(request)
        systemManager.updateSystemProfile(userId, sutId, sname, fname, descr, version, otherSystem, values, Some(files), copySystemParameters, copyStatementParameters).map { _ =>
          ResponseConstructor.constructEmptyResponse
        }
      }
    }.andThen { _ =>
      if (request.body.asMultipartFormData.isDefined) request.body.asMultipartFormData.get.files.foreach { file => FileUtils.deleteQuietly(file.ref) }
    }
  }

  def getSystemById(systemId: Long): Action[AnyContent] = authorizedAction.async { request =>
    authorizationManager.canViewSystem(request, systemId).flatMap { _ =>
      systemManager.getSystemById(systemId).flatMap { system =>
        if (system.isDefined) {
          Future.successful {
            ResponseConstructor.constructJsonResponse(JsonUtil.jsSystem(system.get).toString())
          }
        } else {
          communityLabelManager.getLabel(request, models.Enums.LabelType.System).map { systemLabel =>
            throw NotFoundException(ErrorCodes.SYSTEM_NOT_FOUND, systemLabel + " with ID '" + systemId + "' not found.")
          }
        }
      }
    }
  }

  /**
   * Gets the system profile for the specific system
   */
  def getSystemProfile(sutId:Long): Action[AnyContent] = authorizedAction.async { request =>
    authorizationManager.canViewSystem(request, sutId).flatMap { _ =>
      systemManager.getSystemProfile(sutId).map { system =>
        val json:String = JsonUtil.serializeSystem(system)
        ResponseConstructor.constructJsonResponse(json)
      }
    }
  }

  /**
   * Defines conformance statement for the system
   */
  def defineConformanceStatements(systemId:Long): Action[AnyContent] = authorizedAction.async { request =>
    authorizationManager.canCreateConformanceStatement(request, systemId).flatMap { _ =>
      val actorIds = ParameterExtractor.extractLongIdsBodyParameter(request)
      systemManager.defineConformanceStatements(systemId, actorIds.getOrElse(List.empty)).map { _ =>
        ResponseConstructor.constructEmptyResponse
      }
    }
  }

	def deleteConformanceStatement(sutId: Long): Action[AnyContent] = authorizedAction.async { request =>
    val actorIds = ParameterExtractor.extractLongIdsQueryParameter(request)
    authorizationManager.canDeleteConformanceStatement(request, sutId, actorIds).flatMap { _ =>
      actorIds match {
        case Some(actorIds) =>
          if (actorIds.isEmpty) {
            Future.successful {
              ResponseConstructor.constructBadRequestResponse(ErrorCodes.MISSING_PARAMS, "'ids' parameter should be non-empty")
            }
          } else {
            systemManager.deleteConformanceStatementsWrapper(sutId, actorIds).map { _ =>
              ResponseConstructor.constructEmptyResponse
            }
          }
        case None =>
          Future.successful {
            ResponseConstructor.constructBadRequestResponse(ErrorCodes.MISSING_PARAMS, "'ids' parameter is missing")
          }
      }
		}
	}

  def downloadEndpointConfigurationFile(): Action[AnyContent] = authorizedAction.async { request =>
    val systemId = ParameterExtractor.requiredQueryParameter(request, Parameters.SYSTEM_ID).toLong
    authorizationManager.canViewEndpointConfigurationsForSystem(request, systemId).flatMap { _ =>
      val parameterId = ParameterExtractor.requiredQueryParameter(request, Parameters.PARAMETER_ID).toLong
      Future.successful {
        repositoryUtils.getStatementParameterFile(parameterId, systemId)
      }.map { file =>
        if (file.exists()) {
          Ok.sendFile(
            content = file,
            inline = false
          )
        } else {
          ResponseConstructor.constructNotFoundResponse(ErrorCodes.INVALID_PARAM, "Parameter was not found")
        }
      }
    }
  }

  def getSystemsByOrganization(): Action[AnyContent] = authorizedAction.async { request =>
    val orgId = ParameterExtractor.requiredQueryParameter(request, Parameters.ORGANIZATION_ID).toLong
    val snapshotId = ParameterExtractor.optionalLongQueryParameter(request, Parameters.SNAPSHOT)
    authorizationManager.canViewSystems(request, orgId, snapshotId).flatMap { _ =>
      systemManager.getSystemsByOrganization(orgId, snapshotId).flatMap { list =>
        val checkIfHasTests = ParameterExtractor.optionalBooleanQueryParameter(request, Parameters.CHECK_HAS_TESTS)
        val systemsWithTests = if (checkIfHasTests.isDefined && checkIfHasTests.get) {
          systemManager.checkIfSystemsHaveTests(list.map(x => x.id).toSet).map { hasTests =>
            Some(hasTests)
          }
        } else {
          Future.successful(None)
        }
        systemsWithTests.map { results =>
          val json:String = JsonUtil.jsSystems(list, results).toString
          ResponseConstructor.constructJsonResponse(json)
        }
      }
    }
  }

  def getSystems(): Action[AnyContent] = authorizedAction.async { request =>
    val systemIds = ParameterExtractor.extractLongIdsQueryParameter(request)
    authorizationManager.canViewSystemsById(request, systemIds).flatMap { _ =>
      systemManager.getSystems(systemIds).map { systems =>
        val json = JsonUtil.jsSystems(systems).toString()
        ResponseConstructor.constructJsonResponse(json)
      }
    }
  }

  def searchSystems(): Action[AnyContent] = authorizedAction.async { request =>
    val snapshotId = ParameterExtractor.optionalLongBodyParameter(request, Parameters.SNAPSHOT)
    authorizationManager.canViewSystemsById(request, None, snapshotId).flatMap { _ =>
      val communityIds = ParameterExtractor.extractLongIdsBodyParameter(request, Parameters.COMMUNITY_IDS)
      val organisationIds = ParameterExtractor.extractLongIdsBodyParameter(request, Parameters.ORG_IDS)
      systemManager.searchSystems(communityIds, organisationIds, snapshotId).map { systems =>
        val json = JsonUtil.jsSystems(systems).toString()
        ResponseConstructor.constructJsonResponse(json)
      }
    }
  }

  def searchSystemsInCommunity(): Action[AnyContent] = authorizedAction.async { request =>
    val communityId = ParameterExtractor.requiredBodyParameter(request, Parameters.COMMUNITY_ID).toLong
    val snapshotId = ParameterExtractor.optionalLongBodyParameter(request, Parameters.SNAPSHOT)
    authorizationManager.canViewSystemsByCommunityId(request, communityId, snapshotId).flatMap { _ =>
      val organisationIds = ParameterExtractor.extractLongIdsBodyParameter(request, Parameters.ORG_IDS)
      systemManager.searchSystems(Some(List(communityId)), organisationIds, snapshotId).map { systems =>
        val json = JsonUtil.jsSystems(systems).toString()
        ResponseConstructor.constructJsonResponse(json)
      }
    }
  }

  def getSystemParameterValues(systemId: Long): Action[AnyContent] = authorizedAction.async { request =>
    authorizationManager.canViewSystemsById(request, Some(List(systemId))).flatMap { _ =>
      val onlySimple = ParameterExtractor.optionalBooleanQueryParameter(request, Parameters.SIMPLE)
      systemManager.getSystemParameterValues(systemId, onlySimple).map { values =>
        val json: String = JsonUtil.jsSystemParametersWithValues(values, includeValues = true).toString
        ResponseConstructor.constructJsonResponse(json)
      }
    }
  }

  def checkSystemParameterValues(systemId: Long): Action[AnyContent] = authorizedAction.async { request =>
    authorizationManager.canViewSystemsById(request, Some(List(systemId))).flatMap { _ =>
      systemManager.getSystemParameterValues(systemId).map { values =>
        val valuesWithValidPrerequisites = PrerequisiteUtil.withValidPrerequisites(values)
        val json: String = JsonUtil.jsSystemParametersWithValues(valuesWithValidPrerequisites, includeValues = false).toString
        ResponseConstructor.constructJsonResponse(json)
      }
    }
  }

  def downloadSystemParameterFile(systemId: Long, parameterId: Long): Action[AnyContent] = authorizedAction.async { request =>
    authorizationManager.canViewSystemsById(request, Some(List(systemId))).map { _ =>
      val file = repositoryUtils.getSystemPropertyFile(parameterId, systemId)
      if (file.exists()) {
        Ok.sendFile(
          content = file,
          inline = false
        )
      } else {
        ResponseConstructor.constructNotFoundResponse(ErrorCodes.INVALID_PARAM, "Property was not found")
      }
    }
  }

  def updateSystemApiKey(systemId: Long): Action[AnyContent] = authorizedAction.async { request =>
    authorizationManager.canUpdateSystemApiKey(request, systemId).flatMap { _ =>
      systemManager.updateSystemApiKey(systemId).map { newApiKey =>
        ResponseConstructor.constructStringResponse(newApiKey)
      }
    }
  }

  def ownSystemHasTests(systemId: Long): Action[AnyContent] = authorizedAction.async { request =>
    authorizationManager.canViewSystem(request, systemId).flatMap { _ =>
      testResultManager.testSessionsExistForSystem(systemId).map { hasTests =>
        ResponseConstructor.constructJsonResponse(Json.obj("hasTests" -> hasTests).toString())
      }
    }
  }

}

File: gitb-ui/ui/src/app/services/system.service.ts
/*
 * Copyright (C) 2025 European Union
 *
 * Licensed under the EUPL, Version 1.2 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence"); You may not use this work except in compliance with the Licence.
 *
 * You may obtain a copy of the Licence at:
 *
 * https://interoperable-europe.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on an
 * "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the Licence for
 * the specific language governing permissions and limitations under the Licence.
 */

import { Injectable } from '@angular/core';
import { ROUTES } from '../common/global';
import { FileParam } from '../types/file-param.type';
import { System } from '../types/system';
import { SystemParameter } from '../types/system-parameter';
import { DataService } from './data.service';
import { RestService } from './rest.service';

@Injectable({
  providedIn: 'root'
})
export class SystemService {

  constructor(
    private readonly restService: RestService,
    private readonly dataService: DataService
  ) { }

  getSystemsByOrganisation(orgId: number, snapshotId?: number, checkIfHasTests?: boolean) {
    let params: any = {
      organization_id: orgId
    }
    if (snapshotId !== undefined) {
      params.snapshot = snapshotId
    }
    if (checkIfHasTests !== undefined) {
      params.check_has_tests = checkIfHasTests
    }
    return this.restService.get<System[]>({
      path: ROUTES.controllers.SystemService.getSystemsByOrganization().url,
      authenticate: true,
      params: params
    })
  }

  getSystemById(systemId: number) {
    return this.restService.get<System>({
      path: ROUTES.controllers.SystemService.getSystemById(systemId).url,
      authenticate: true
    })
  }

  getSystem(systemId: number) {
    return this.restService.get<System>({
      path: ROUTES.controllers.SystemService.getSystemProfile(systemId).url,
      authenticate: true
    })
  }

  searchSystems(communityIds: number[]|undefined, organisationIds: number[]|undefined, snapshotId?: number) {
		const data: any = {}
		if (communityIds && communityIds.length > 0) {
		  data["community_ids"] = communityIds.join(',')
		}
		if (organisationIds && organisationIds.length > 0) {
		  data["organization_ids"] = organisationIds.join(',')
		}
    if (snapshotId != undefined) {
      data["snapshot"] = snapshotId
    }
    return this.restService.post<System[]>({
      path: ROUTES.controllers.SystemService.searchSystems().url,
      authenticate: true,
      data: data
    })
  }

  searchSystemsInCommunity(communityId: number, organisationIds: number[]|undefined, snapshotId?: number) {
		const data: any = {
      community_id: communityId
    }
		if (organisationIds && organisationIds.length > 0) {
		  data["organization_ids"] = organisationIds.join(',')
		}
    if (snapshotId != undefined) {
      data['snapshot'] = snapshotId
    }
    return this.restService.post<System[]>({
      path: ROUTES.controllers.SystemService.searchSystemsInCommunity().url,
      authenticate: true,
      data: data
    })
  }

  getSystems(systemIds?: number[]) {
    let params: any = {}
    if (systemIds !== undefined && systemIds.length > 0) {
      params.ids = systemIds.join(',')
    }
    return this.restService.get<System[]>({
      path: ROUTES.controllers.SystemService.getSystems().url,
      authenticate: true,
      params: params
    })
  }

  getSystemParameterValues(systemId: number, onlySimple?: boolean) {
    let params = undefined
    if (onlySimple != undefined) {
      params = {
        simple: onlySimple
      }
    }
    return this.restService.get<SystemParameter[]>({
      path: ROUTES.controllers.SystemService.getSystemParameterValues(systemId).url,
      authenticate: true,
      params: params
    })
  }

  downloadSystemParameterFile(systemId: number, parameterId: number) {
		return this.restService.get<ArrayBuffer>({
			path: ROUTES.controllers.SystemService.downloadSystemParameterFile(systemId, parameterId).url,
			authenticate: true,
			arrayBuffer: true
		})
  }

  updateSystem(systemId: number, sname: string, fname: string, description: string|undefined, version: string|undefined, organisationId: number, otherSystem: number|undefined, processProperties: boolean, properties: SystemParameter[], copySystemParameters: boolean, copyStatementParameters: boolean) {
    const data: any = {
      system_sname: sname,
      system_fname: fname
    }
    if (version != undefined) {
      data.system_version = version
    }
    if (description != undefined) {
      data.system_description = description
    }
    if (otherSystem) {
      data.other_system = otherSystem
      data.sys_params = copySystemParameters
      data.stm_params = copyStatementParameters
    }
    let files: FileParam[]|undefined
    if (processProperties) {
      const props = this.dataService.customPropertiesForPost(properties)
      data.properties = props.parameterJson
      files = props.files
    }
    data.organization_id = organisationId
    return this.restService.post<void>({
      path: ROUTES.controllers.SystemService.updateSystemProfile(systemId).url,
      data: data,
      files: files,
      authenticate: true
    })
  }

  registerSystemWithOrganisation(sname: string, fname: string, description: string|undefined, version: string|undefined, orgId: number, otherSystem: number|undefined, processProperties: boolean, properties: SystemParameter[], copySystemParameters: boolean, copyStatementParameters: boolean) {
    const data: any = {
      system_sname: sname,
      system_fname: fname,
      organization_id: orgId
    }
    if (version != undefined) {
      data.system_version = version
    }
    if (otherSystem) {
      data.other_system = otherSystem
      data.sys_params = copySystemParameters
      data.stm_params = copyStatementParameters
    }
    if (description != undefined) {
      data.system_description = description
    }
    let files: FileParam[]|undefined
    if (processProperties) {
      const props = this.dataService.customPropertiesForPost(properties)
      data.properties = props.parameterJson
      files = props.files
    }
    return this.restService.post<void>({
      path: ROUTES.controllers.SystemService.registerSystemWithOrganization().url,
      data: data,
      files: files,
      authenticate: true
    })
  }

  deleteSystem(systemId: number, organisationId: number) {
    return this.restService.delete<void>({
      path: ROUTES.controllers.SystemService.deleteSystem(systemId).url,
      params: {
        organization_id: organisationId
      },
      authenticate: true
    })
  }

  defineConformanceStatements(system: number, actorIds: Set<number>) {
    return this.restService.post<void>({
      path: ROUTES.controllers.SystemService.defineConformanceStatements(system).url,
      data: {
        ids: [...actorIds].join(',')
      },
      authenticate: true
    })
  }

  checkSystemParameterValues(systemId: number) {
    return this.restService.get<SystemParameter[]>({
      path: ROUTES.controllers.SystemService.checkSystemParameterValues(systemId).url,
      authenticate: true
    })
  }

  deleteConformanceStatement(systemId: number, actorIds: number[]) {
    return this.restService.delete<void>({
      path: ROUTES.controllers.SystemService.deleteConformanceStatement(systemId).url,
      authenticate: true,
      params: {
        ids: actorIds.join(',')
      }
    })
  }

  downloadEndpointConfigurationFile(systemId: number, parameterId: number) {
		return this.restService.get<ArrayBuffer>({
			path: ROUTES.controllers.SystemService.downloadEndpointConfigurationFile().url,
			authenticate: true,
      params: {
        system_id: systemId,
        parameter_id: parameterId
      },
			arrayBuffer: true
		})
  }

  updateSystemApiKey(systemId: number) {
    return this.restService.post<string>({
      path: ROUTES.controllers.SystemService.updateSystemApiKey(systemId).url,
      authenticate: true,
      text: true
    })
  }

  ownSystemHasTests(systemId: number) {
    return this.restService.get<{hasTests: boolean}>({
      path: ROUTES.controllers.SystemService.ownSystemHasTests(systemId).url,
      authenticate: true
    })
  }

}
